pkg/tlsx/tls/tls.go

```diff
@@ -116,12 +117,14 @@ func (c *Client) ConnectWithOptions(hostname, ip, port string, options clients.C
 	if rawConn == nil {
 		return nil, fmt.Errorf("could not connect to %s", address)
 	}
+
 	var resolvedIP string
-	if !iputil.IsIP(hostname) {
+	if ip != "" {
+		resolvedIP = ip
+	} else if iputil.IsIP(hostname) {
+		resolvedIP = hostname
```

Comment:
Instead of handling edgecase where `ip` is passed in `hostname` field . How about verifying at beginning of function call if 

hostname is indeed a hostname and not ip
ip is valid and not a hostname 

this might solve future errors . current issue with nuclei is that we are passing hostname in both fields (ip & hostname) https://github.com/projectdiscovery/nuclei/blob/05d607642a11449262fba8d079025dc5325dec8d/v2/pkg/protocols/ssl/ssl.go#L164-L167

Reply:
@tarunKoyalwar, the `ip` parameter is only used in case of custom ip override, typically, the hostname might also contain a DNS name, and it defers to fastdialer to connect to the first responsive ip. I think here the check is legitimate as we are handling the cases:
- Custom ip provided => use that
- input was an ip => use that
- input was a hostname => retrieve the dialed ip from fastdialer

This should be addressed uniformly with https://github.com/projectdiscovery/metainput across all the tools (currently prototyped in nuclei only), as various parameters will change to a unique struct definition.

Reply:
Thanks , Got it